daffodil-lib/src/main/scala/org/apache/daffodil/api/Validator.scala

```diff
@@ -54,20 +55,31 @@ trait ValidatorFactory {
 
 /**
  * Results of a validation execution
- * @param warnings [[org.apache.daffodil.api.ValidationWarning]] objects
- * @param errors [[org.apache.daffodil.api.ValidationFailure]] objects
  */
-final case class ValidationResult(warnings: java.util.Collection[ValidationWarning], errors: java.util.Collection[ValidationFailure])
+trait ValidationResult {
+  def warnings(): java.util.Collection[ValidationWarning]
+  def errors(): java.util.Collection[ValidationFailure]
+}
+
+/**
+ * Provider of raw validation output
+ */
+trait RawValidationResult {
+  def rawValidationData(): Array[Byte]
+}
 
 object ValidationResult {
   /**
    * an empty [[org.apache.daffodil.api.ValidationResult]]
    */
   val empty: ValidationResult = ValidationResult(Seq.empty, Seq.empty)
 
-  def apply(warnings: Seq[ValidationWarning], errors: Seq[ValidationFailure]): ValidationResult = {
+  def apply(w: Seq[ValidationWarning], e: Seq[ValidationFailure]): ValidationResult = {
     import scala.collection.JavaConverters.asJavaCollectionConverter
-    ValidationResult(warnings.asJavaCollection, errors.asJavaCollection)
+    new ValidationResult{
+      val warnings: java.util.Collection[ValidationWarning] = w.asJavaCollection
+      val errors: java.util.Collection[ValidationFailure] = e.asJavaCollection
+    }
```

Comment:
Here's an interesting tidbit.  No other place in the Daffodil codebase uses java.util.Collection or the asJavaCollection converter from Seq to Collection.  I'm also surprised that an API class in daffodil-lib would use Collection instead of Seq, although doing so lets Java callers use ValidationResult directly.  If all this is a new and better practice, I suspect we could apply it in more places.

Reply:
Good point.  IIRC this path was taken for ease of transition in the JAPI, and it was done in the original refactor PR, not new here.

But looking back now it probably should be a scala type and converted in the JAPI.  Ill see what a refactor here looks like.

Reply:
I did a refactor on this and it looks good, but grows pretty large considering the changes needed to keep the internal API object out of the external APIs.

Going to follow up with a ticket like "Refactor Validation API to remove Java types"